src/NUnitFramework/framework/Internal/Execution/TextMessageWriter.cs

```diff
@@ -296,6 +294,18 @@ private void WriteExpectedLine(object expected, Tolerance tolerance)
         private void WriteActualLine(ConstraintResult result)
         {
             Write(Pfx_Actual);
+
+            // Write the actual item count, when ActualValue is a collection
+            if (result.Name == "ExactCount" && result.ActualValue is IEnumerable)
```

Comment:
So that every possible constraint doesn't end up putting code in this shared location, we use custom constraint results for this type of thing. In other words, you'd create a private sealed class `ExactCountConstraint.ExactCountConstraintResult`, override `WriteActualValueTo`, and use it [here](https://github.com/nunit/nunit/blob/d411123f1c705cc7ad147c67d2420d97b0706626/src/NUnitFramework/framework/Constraints/ExactCountConstraint.cs#L86). Have you tried that?

Reply:
@TobiasSekan Emphasizing @jnm2 's point...

`TextMessageWriter` knows how to put out a message in text form based on what the constraint asked to have done. Constraints know what the `IMessageWriter` interface can do and call the appropriate methods. The message writer doesn't know anything about constraints and shouldn't. It only varies what is put out based on the particular type of output. So, a hypothetical `HtmlMessageWriter` would put put different characters for the same call from what a `TextMessageWriter` does.

A constraint, like `ExactCountConstraint` decides what it's output will look like. If you examine constraint code in NUnit, you will see that many different kinds of output are produced. Some constraints do this by direct use of the message writer, while others do it by creating a custom constraint result, as suggested by @jnm2. A custom result is needed when the output is significantly different from the standard or particularly complex.

I suggest you look at the code for various constraints, especially those that output lists of collection elements to see how we do it and modify the `ExactCountConstraint` output, leaving `TextMessageWriter` as it is.